stream.go

```diff
@@ -226,6 +234,19 @@ func (cs *clientStream) CloseSend() (err error) {
 	return
 }
 
+func (cs *clientStream) finish(err error) {
```

Comment:
On Thu, Jun 25, 2015 at 5:51 PM, David Symonds <notifications@github.com>
wrote:

> In stream.go
> <https://github.com/grpc/grpc-go/pull/235#discussion_r33320833>:
>
> > @@ -226,6 +234,19 @@ func (cs *clientStream) CloseSend() (err error) {
> >  	return
> >  }
> >
> > +func (cs *clientStream) finish(err error) {
>
> Yes, this would break because of that. I couldn't figure out how this code
> should clearly delimit when a clientStream is "done", since both SendMsg
> and RecvMsg appear to indicate the termination of the stream when things
> fail.
>
> Does clientStream need a mutex to mediate this? What's supposed to happen
> if concurrent SendMsg/RecvMsg calls are running?
>

i) add clientStream.mu to protect payload tracing which could happens in
SendMsg and RecvMsg at same time;
ii)  a streaming rpc is always terminated via a stream.Recv() or
stream.CloseAndRecv(0 (both invoke RecvMsg) to get the final status even
though stream.Send(...) could get an IO error first. So trace termination
should be only in RecvMsg.

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc-go/pull/235/files#r33320833>.
>